examples/common/hf_fsdp.py

```diff
@@ -0,0 +1,220 @@
+        - gpt_neox.layers: (GPTNeoXForCausalLM)
+    """
+    hidden_layers_attrs = (
+        'transformer.h',
+        'model.decoder.layers',
+        'gpt_neox.layers',
+        'block',  # T5, BART, Pegasus (from encoder)
+        'layers',  # ProphetNet, Marian (from encoder)
+    )
+    return findattr(model, hidden_layers_attrs)
+
+
+def hf_get_tied_embedding_weights(model: PreTrainedModel):
+    """Returns the embeddings, which are weight tied layers.
+
+    NOTE: Different model configurations have different embedding attribute names.
+        - wte: (GPT2LMHeadModel, GPTJForCausalLM, GPTNeoForCausalLM)
+        - word_embeddings: (BloomForCausalLM)
+        - embed_tokens: (OPTForCausalLM)
+        - GPT NeoX doesn't weight tie
+    """
+    tied_embedding_attrs = (
+        'wte',
+        'word_embeddings',
+        'embed_tokens',
+    )
+    return findattr(model, tied_embedding_attrs)
+
+
+# /end helper functions
+
+
+def prepare_hf_model_for_fsdp(model: PreTrainedModel) -> None:
+    """FSDP wrap a HuggingFace model.
+
+    Call specific functions
+    """
+    if model.config.is_encoder_decoder:
+        prepare_hf_enc_dec_model_for_fsdp(model)
+    else:
+        # many common decoder-only model do not set the flag
+        # model.config.is_decoder, so we can't trust it
+        prepare_hf_causal_lm_model_for_fsdp(model)
+
+
+def prepare_hf_causal_lm_model_for_fsdp(model: PreTrainedModel) -> None:
+    """FSDP wrap a HuggingFace decoder.
+
+    Wrap any model for FSDP which follows one of the 3 existing conventions from
+    HuggingFace for decoder-only LLMs.
+    """
+    causal_base_model = hf_get_causal_base_model(model)
+    model_block = hf_get_hidden_layers(model)  # type: ignore
+    lm_head = hf_get_lm_head(model)
+    tied_embeddings = hf_get_tied_embedding_weights(
+        causal_base_model)  # type: ignore
+    modules = {
+        "base_model": causal_base_model,
+        "model_block": model_block,
+        "lm_head": lm_head,
+        "tied_embeddings": tied_embeddings
+    }
+
+    for mod_name, module in modules.items():
+        if module is None:
+            raise ValueError(
+                f'Unable to FSDP-wrap this model! `{mod_name}` does not ' +
+                'follow common layer/weight naming conventions.')
+    block_type = type(model_block[0])  # type: ignore
+    # When using the HF LM models,
+    # the weights of the self.lm_head and self.transformer.wte are tied.
+    # This tying occurs inside the `self.post_init()` function.
+    # This is a hurdle for FSDP because they need to be in the same FSDP block
+    # These lines ensures that both modules stay together in the top-most block when
+    # the model has this tying enabled (almost all do; this property defaults to True)
+    if model.config.tie_word_embeddings:
+        causal_base_model._fsdp_wrap = False  # type: ignore
+        tied_embeddings._fsdp_wrap = False  # type: ignore
+        lm_head._fsdp_wrap = False  # type: ignore
```

Comment:
If you take out all of this code, and make sure that `fsdp_config.min_params` is really really big, like `1e12`, so that it doesn't get in the way, does everything still work?